jwplatform/upload.py

```diff
@@ -0,0 +1,137 @@
+        self.upload_retry_count = retry_count
+        self.file = file
+        self.client = client
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+    def upload(self):
+
+        if self.target_part_size < constants.MIN_PART_SIZE:
+            raise ValueError(f"The part size has to be at least greater than {constants.MIN_PART_SIZE} bytes.")
+
+        filename = self.file.name
+        file_size = common.get_file_size(self.file)
+        part_count = file_size // self.target_part_size + 1
+
+        if part_count > 1000:
+            raise ValueError(f"The given file cannot be divided into more than 1000 parts. Please try increasing the "
+                             f"target part size.")
+
+        # Get the part links
+        upload_links = self._get_pre_signed_part_links(part_count)
+
+        # Upload the parts
+        for part_number in range(1, part_count + 1):
+            bytes_chunk = self.file.read(self.target_part_size)
+            if part_number < part_count and len(bytes_chunk) != self.target_part_size:
+                raise IOError("Failed to read enough bytes")
+            retry_count = 0
+            while retry_count < self.upload_retry_count:
+                try:
+                    self._upload_part(bytes_chunk, part_number, upload_links)
+                    break
+                except (DataIntegrityError, HTTPError):
+                    self.logger.warning(f"Encountered error upload part {part_number} of {part_count} for file {filename}."
+                                        f" Retrying.")
+                    retry_count = retry_count + 1
+
+            if retry_count >= self.upload_retry_count:
+                raise MaxRetriesExceededError(f"Max retries ({self.upload_retry_count}) exceeded while uploading part"
+                                              f" {part_number} of {part_count} for file {filename}.")
+
+        # Mark upload as complete
+        self._mark_upload_completion()
+
+    def _upload_part(self, bytes_chunk, part_number, upload_links):
+        # Add a S3 server-side checksum validation too if possible.
+        computed_hash = self._compute_part_hash(bytes_chunk)
+
+        # Check if the file has already been uploaded and the hash matches. Return immediately without doing anything
+        # if the hash matches.
+        upload_hash = upload_links[part_number - 1]["etag"] if "etag" in upload_links[part_number - 1] else None
+        if upload_hash:
+            if repr(upload_hash) == repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+                return
+
+        upload_link = upload_links[part_number - 1]["upload_link"] if "upload_link" in upload_links[part_number - 1] \
+            else None
+        if not upload_link:
+            raise Exception(f"Invalid upload link for part {part_number}.")
+
+        resp = requests.put(upload_links[part_number - 1]["upload_link"], data=bytes_chunk)
```

Comment:
I'm not a huge fan of requests in a library like this given the recent changes to pip's dependency resolution. requests is a common package and this would require us to either pin no version (and risk surprising users with breaking changes to requests that we don't handle) or pin a major version in which case someone will always be unhappy we don't support _their_ version.

That said I also don't think requests is providing much value here.

We should probably provide an exception class for S3 upload errors similar to [APIError](https://github.com/jwplayer/jwplatform-py/blob/f12a6211d247606ff6df07e7faba02bfab311716/jwplatform/client.py#L66) but I don't think we need to go to the same effort of parsing the response and providing distinct classes for different types of errors.

Reply:
After reading how the exception here is never actually returned to the user it's _possible_ that last point about an exception class is not necessary.

I don't really feel great about hiding exceptions from the user though. Currently it's impossible for a user to debug and really know what went wrong. Moving the raise of `MaxRetriesExceededError` inside the `except` might be enough to expose the most recent failure to the user.